Sources/SwiftLexicalLookup/SimpleLookupQueries.swift

```diff
@@ -0,0 +1,227 @@
+    }
+  }
+
+  // MARK: - walkParentTree helper methods
+
+  /// Callect the first syntax node matching the collection type up to a function boundary.
+  fileprivate func walkParentTreeUpToFunctionBoundary<T: SyntaxProtocol>(
+    at syntax: Syntax?,
+    collect: T.Type
+  ) -> T? {
+    walkParentTreeUpToFunctionBoundary(at: syntax, collect: collect, stopWithFirstMatch: true).first
+  }
+
+  /// Callect syntax nodes matching the collection type up to a function boundary.
+  fileprivate func walkParentTreeUpToFunctionBoundary<T: SyntaxProtocol>(
+    at syntax: Syntax?,
+    collect: T.Type,
+    stopWithFirstMatch: Bool = false
+  ) -> [T] {
+    walkParentTree(
+      upTo: [
+        MemberBlockSyntax.self,
+        FunctionDeclSyntax.self,
+        InitializerDeclSyntax.self,
+        DeinitializerDeclSyntax.self,
+        AccessorDeclSyntax.self,
+        ClosureExprSyntax.self,
+      ],
+      collect: collect,
+      stopWithFirstMatch: stopWithFirstMatch
+    )
+  }
+
+  /// Callect syntax nodes matching the collection type up until encountering one of the specified syntax nodes.
+  private func walkParentTree<T: SyntaxProtocol>(
+    upTo stopAt: [SyntaxProtocol.Type],
+    collect: T.Type,
+    stopWithFirstMatch: Bool = false
+  ) -> [T] {
+    var matches: [T] = []
+    var nextSyntax: Syntax? = Syntax(self)
+    while let currentSyntax = nextSyntax {
+      if stopAt.contains(where: { currentSyntax.is($0) }) {
+        break
+      }
+
+      if let matchedSyntax = currentSyntax.as(T.self) {
+        matches.append(matchedSyntax)
+        if stopWithFirstMatch {
+          break
+        }
+      }
+
+      nextSyntax = currentSyntax.parent
+    }
+
+    return matches
+  }
+}
+
+extension FallThroughStmtSyntax {
+  /// Returns the source and destination of a `fallthrough`.
+  ///
+  /// - Returns: `source` as the switch case that encapsulates the `fallthrough` keyword and `destination` as the switch case that the `fallthrough` directs to.
+  ///
+  /// Example usage:
+  /// ```swift
+  /// switch value {
+  /// case 2:
+  ///   doSomething()
+  ///   fallthrough
+  /// case 1:
+  ///   doSomethingElse()
+  /// default:
+  ///   break
+  /// }
+  /// ```
+  /// When calling this function at the `fallthrough`, it returns `case 2` and `case 1` in this exact order. The `nil` results handle ill-formed code: there's no `source` if the `fallthrough` is outside of a case. There's no `destination` if there is no case or `default` after the source case.
+  ///
+  @_spi(Experimental) public func lookupFallthroughSourceAndDest()
+    -> (source: SwitchCaseSyntax?, destination: SwitchCaseSyntax?)
+  {
+    return lookupFallthroughSourceAndDestination(at: self)
+  }
+
+  // MARK: - lookupFallthroughSourceAndDest
+
+  /// Given syntax node position, returns the current switch case and it's fallthrough destination.
+  private func lookupFallthroughSourceAndDestination(
+    at syntax: SyntaxProtocol
```

Comment:
Similar to other comments I have, I think having a function that takes a syntax node can be confusing. Personally, I would just implement this in `lookupFallthroughSourceAndDestination` that doesn’t take any parameters.